src/librbd/cache/pwl/LogEntry.h

```diff
@@ -173,17 +191,31 @@ class WriteLogEntry : public GenericWriteLogEntry {
   #ifdef WITH_RBD_RWL
   void init_pmem_buffer(std::vector<WriteBufferAllocation>::iterator allocation);
   #endif
+  void init_ssd_bl(bufferlist &src_bl, uint64_t off, uint64_t len);
```

Comment:
Nit: it would be clearer if we had derived versions of `WriteLogEntry` for SSD vs RWL so we don't need to invoke different methods depending on the current version. Instead, at construction time, the SSD cache will create the SSD  version and the RWL cache will create the RWL version.

Reply:
Current overall class hierarchy design wouldn't allow that to happen easily - `Request` class is responsible for creating the `LogOperation` objects which in turn create `LogEntry` objects. And right now `Request` class takes the `AbstractWriteLog` as the template type, so there isn't way to template specialize for `SSDWriteLog` and `ReplicatedWriteLog`. Multiple inheritance would allow that though. But I'd like to note that, current changes should mostly address the differences for SSD cache vs RWL in `LogEntry`,`LogOperation`, and `Request` classes and I don't anticipate many more changes that are specific to SSD in those classes.

Reply:
You could use a builder or `AbstractWriteLog` could be templated on both `ImageCtx` and the cache type so that you can then use template specialization to create a `type_traits`-like helper class for both RWL vs SSD and allow the correct derived types to be utilized. I think that would be more maintainable in the future vs trying to push the specializations down to the base classes.

Reply:
Done. Used the Builder approach.